torch/_higher_order_ops/out_dtype.py

```diff
@@ -0,0 +1,151 @@
+
+    flat_inputs = pytree.tree_flatten(args)[0] + [torch.ones(1, dtype=output_dtype)]
+    promote_dtype: torch.dtype = elementwise_dtypes(
+        *flat_inputs,
+        type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+    )[0]
+
+    casted_args = pytree.tree_map_only(
+        torch.Tensor, lambda arg: arg.to(dtype=promote_dtype), args
+    )
+    res = op(*casted_args).to(dtype=output_dtype)
+    return res
+
+
+@out_dtype.py_impl(DispatchKey.Autograd)
+def out_dtype_autograd(
+    op: torch._ops.OpOverload,
+    output_dtype: torch.dtype,
+    *args
+):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("out_dtype's first argument must be an OpOverload")
+
+    # TODO: support autograd
+    flat_operands, _ = pytree.tree_flatten(args)
+    assert all(
+        not f.requires_grad for f in flat_operands if isinstance(f, torch.Tensor)
+    ), "Autograd is not supported for out_dtype"
+
+    _ = torch._C.ExcludeDispatchKeyGuard(
+        torch._C.DispatchKeySet(torch._C.DispatchKey.AutogradCPU)
+    )
+    return out_dtype(op, output_dtype, *args)
+
+
+@out_dtype.py_impl(ProxyTorchDispatchMode)
+def out_dtype_proxy(
+    op: torch._ops.OpOverload,
+    output_dtype: torch.dtype,
+    *args
+):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("out_dtype's first argument must be an OpOverload")
+
+    mode = _get_current_dispatch_mode()
+    assert (mode is not None), "Mode should always be enabled for python fallback key"
+    with _pop_mode_temporarily() as mode:
+        if mode.enable_tracing:
+            return trace_out_dtype(mode, out_dtype, op, output_dtype, *args)
+        else:
+            return out_dtype(op, output_dtype, *args)
+
+
+@out_dtype.py_impl(FakeTensorMode)
+def out_dtype_fake_tensor_mode(
+    op: torch._ops.OpOverload,
+    output_dtype: torch.dtype,
+    *args
+):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("out_dtype's first argument must be an OpOverload")
+    return out_dtype_dense(op, output_dtype, *args)
+
+
+@out_dtype.py_impl(torch._C.DispatchKey.Functionalize)
+def out_dtype_func1(op, output_dtype, *args):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("out_dtype's first argument must be an OpOverload")
+    if op._schema.is_mutable:
+        raise ValueError("out_dtype's first argument needs to be a functional operator")
+
+    reapply_views = torch._C._functionalization_reapply_views_tls()
+    # At this point, we will see functionalized tensors, so need to unwrap them first
+    unwrapped_args = tuple(
+        _unwrap_all_tensors_from_functional(arg, reapply_views=reapply_views)
+        for arg in args
+    )
+    # pyre-ignore
+    with torch._C._ExcludeDispatchKeyGuard(torch._C.DispatchKeySet(torch._C.DispatchKey.Functionalize)):
+        res = out_dtype(op, output_dtype, *unwrapped_args)
+    return _wrap_all_tensors_to_functional(res, level=0)
+
```

Comment:
You're going to want to copy the logic from here: https://github.com/pytorch/pytorch/blob/231364fd064e08220fa78b447a817db8cddb1b28/aten/src/ATen/FunctionalizeFallbackKernel.cpp#L50

1. For each Tensor input, we need to: call sync_ on it, then use from_functional_tensor to unwrap the Tensor
2. Then, we need to do the ExcludeDispatchKeyGuard
3. Finally, we should rewrap everything in FunctionalTensorWrapper

Your code does 2 and 3, I think it is missing the sync_

Reply:
I call _unwrap_all_tensors_from_functional beforehand, which I think calls sync_ [here](https://www.internalfb.com/code/fbsource/[eee9a617ea03]/fbcode/caffe2/torch/_functorch/eager_transforms.py?lines=13160). Is that enough?

Reply:
Yes, you're right -- sorry I missed that. This looks good to me then.

Reply:
what does sync do?